proposals/route-blinding.md

```diff
@@ -0,0 +1,242 @@
+compute the following:
+
+```text
+  ss(0) = H(k(0) * E(0))
+  rho(0) = HMAC256("rho", ss(0))
+  E(1) = H(E(0) || ss(0)) * E(0)
+```
+
+It uses `rho(0)` to decrypt the `encrypted_blob(0)` and discover the `scid` to forward to.
+It forwards the onion to the next node and includes `E(1)` in a TLV field in the message
+extension.
+
+All the following intermediate nodes `N(i)` do the following steps:
+
+```text
+  E(i) <- extracted from TLV extension
+  ss(i) = H(k(i) * E(i))
+  b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)
+  Use b(i) to decrypt the incoming onion
+  rho(i) = HMAC256("rho", ss(i))
+  Use rho(i) to decrypt the `encrypted_blob` inside the onion and discover the next node
+  E(i+1) = H(E(i) || ss(i)) * E(i)
+  Forward the onion to the next node and include E(i+1) in a TLV field in the message extension
+```
+
+### Receiving from a blinded route
+
+When `N(r)` receives the onion message and `E(r)` in the TLV extension, she does the same
+unwrapping as intermediate nodes. The difference is that the onion will be a final onion.
+
+### Unblinding channels via fee probing
+
+The fees and cltv for the blinded route can be abused by the sender to try to unblind the real
+nodes and channels used. The sender can create onions with increased fees/cltv for the first
+blinded hop, starting with very low values. While the fee/cltv is below the real fee of the first
+hop, the sender will get an error from `N(0)`. Once the fee/cltv proposed actually satisfies the
+first hop's requirements, the error will come from another node `N(i)` inside the blinded path.
+
+The sender can then unblind channels one-by-one by discovering their real fees/cltv and matching
+those to existing channels in the graph.
+
+To mitigate this, when nodes along the blinded path are offered an invalid HTLC, they should:
+
+* return a dummy error encrypted with a throw-away key: the sender will receive an error she can't
+  decrypt and doesn't know which node generated it
+* hold the HTLC for a random amount of time before sending the error (otherwise the sender can
+  still use timing to guess what node errored out)
+
+Even with such mitigations the sender can discover the real fees/cltv of one of the blinded
+channels. To do so she uses the correct fees and cltv for all but one channel, and for that target
+channel she tries fees/cltv until the payment succeeds. Once the payment succeeds she knows the
+approximate fees/cltv of the target channel (but since the payment succeeded, she can't continue
+probing).
+
+Maybe nodes along the blinded path could use slightly different fees/cltv than what they publicly
+advertize? Or the recipient could add some fuzzing to it to blind them more?
+
+Are those mitigations enough? Or can a clever attacker still work around them?
+
+## Tips and Tricks
+
+### Dummy hops
+
+The sender knows an upper bound on the distance between the recipient and `N(0)`. If the recipient
```

Comment:
To make dummy hops plausible you need fees to be realistic, which means amount paid by sender is going to be artificially inflated. Assuming fees are uniform to reach multiple endpoints of the network, this may be a deanonymization heuristic.

Let's say you want to pay 1000 sat. It costs you 1100 for a one-hop distance, 1200 for a two-hop distance, ..., 1600 for a six-hop distance, using normal routing. 

Now same sender wants to pay 1000 sat using a six-hops-sized blinded_route and after fees addition, total cost is 1430.  Compared to the normal routing data, you may observe a discrepancy which reveals than the last 2-3 hops are dummy ones.

So making realistic dummy hops means to align with network data, which is a hard-requirement for a low-resource node, and hard to implement right in a dummy_hops fees estimation algorithm.

Is this a realistic heuristic ?

Reply:
> To make dummy hops plausible you need fees to be realistic, which means amount paid by sender is going to be artificially inflated.

That's a good point, but then when using blinded paths or rendezvous I believe the price of the item you're selling should be the price that reaches the introduction point (recipient pays the blinded path). That should remove this issue, doesn't it?

Reply:
Okay, it's less on item price that the fact you need network data to make dummy_hops fees plausible compare to a fully-known payment path for the same value routed. I need to think about this more..

Reply:
Since you need to know the fees along the blinded path, you can select the same feerate as those (real) channels for your dummy hop.